Update and give access to invisible automated testing page doc (#136)
SuffolkLITLab/docassemble-AssemblyLine-documentation: docs/automated_testing.md

---
id: automated_testing
title: Automated Integrated Testing
sidebar_label: Automated Testing
slug: /automated_integrated_testing
---

<!-- original: https://docs.google.com/document/d/1hkNr78mrU3Ha98tPUL4OKWi3eNnt-1Sba7L8470u06g/edit# -->

🚧 Reference for writing automated integrated tests with the Assembly Line testing framework.

<!-- ## Setup

For now, talk to us about it.

- Have admin permissions on the repo
- Use the setup interview (https://apps-dev.suffolklitlab.org/start/test-setup/)
- Add this element to the page somewhere: <div data-variable="${ encode_name(str( user_info().variable )) }" id="trigger" aria-hidden="true" style="display: none;"></div>

Secrets

PLAYGROUND_PASSWORD
PLAYGROUND_ID

> PLAYGROUND_EMAIL

Author: This info is actually not 'your' info, it's the info of the testing account. This info is only needed for local development. Online, the setup interview walks people through putting in the right info.

> Secrets
> 
> PLAYGROUND_PASSWORD
> PLAYGROUND_ID
> PLAYGROUND_EMAIL

Author: ```suggestion
If you run the tests locally, add these environment variables to your `.env` file:

 -` PLAYGROUND_EMAIL`: the docassemble server email for the account that is running the tests.
- `PLAYGROUND_PASSWORD`: the docassemble server password for the account that is running the tests.
- `PLAYGROUND_ID`: the id that shows up in interviews when the testing account runs the tests from the server playground.
```

> PLAYGROUND_EMAIL

Reviewer: Thanks for the correction, and exactly the reason this short clarifier would help a lot, including me! :sweat_smile:

-->

## Summary

> 1. You write and edit `.feature` test files in your Sources folder.

Reviewer: ```suggestion
1. You write and edit `.feature` test files in your Sources folder that contain a series of steps that the test should take when filling out the interview.
```

Author: I really want this to be a very bare bones summary. A very quick reference. How about this?

```suggestion
1. You write and edit `.feature` test files in your Sources folder. Instructions for writing tests are below.
```

Reviewer: I guess, but the issue I had with this line is that if you don't already know what a `.feature` file is, it tells you very little. Just knowing that it's some steps for the tester is good.

But this is minor for just the deep-dive version of the code. Feel free to just ignore for now.

Author: I think this is something that the documentation guidelines site might help with - figuring out where different pieces might go.

> 1. There's a maximum of 2 minutes per individual Step or table row. Longer than that and the test will error. That can't be customized yet.
> 1. Tests are run when you commit.

Reviewer: ```suggestion
1. Tests are run when you commit (there's a maximum of 2 minutes per individual Step or table row. Longer than that and the test will error. That can't be customized yet).
```

Author: I'd like each item to be as short and well-contained as possible. I think the two concepts are distinct and unrelated. I only mention the timeout because I think it's pretty crucial. How about just switching the order?

```suggestion
1. Tests are run when you commit.
1. There's a maximum of 2 minutes per individual Step or table row. Longer than that and the test will error. That can't be customized yet.
```

Reviewer: Taking a step back I guess, what are these items? They read more like steps that the test author takes, but the "There's a maximum of 2 minutes per individual Step..." item isn't something you can do. The second item probably doesn't need to be mentioned here, and can be moved later, even if it's crucial (i.e., is it information that someone, encountering the testing system for the first time, would even remember by the time they got to writing tests?)

Author: Ah, I see. Can we leave it this way for the MVP and revisit the best place to put this info when re-organizing a la divio? If so, resolve for now and we can make an issue for it to avoid forgetting.

Author: #138

  @fast @child_benefits
  Scenario: child has benefits
    Given I start the interview at "some_file_name"
    And I get to the question id "benefits" with this data:
      | var | value | trigger |
      | x[i].name.first | Cashmere | children[0].name.first |
      | x[i].name.first | Casey | children[1].name.first |
      | x.there_are_any | True | children.there_are_any |
      | x.there_is_another | 2 | children.there_is_another |
    When I set the var "benefits['SSI']" to "true"
    And I tap to continue
    Then the question id should be "download"
    And I download "some_motion.pdf"

```

## First test

You can write a really simple test right away that just makes sure your a file runs using the name of the file. Write a `Scenario` for each file you want to test.

```
Feature: Interviews load

  Scenario: The 209A loads
    Given I start the interview at "ma_209a_package"

  Scenario: The Plaintiff's Motion to Modify loads
    Given I start the interview at "plaintiffs_motion_to_modify_209a"
```

More complex tests might wait till your code is pretty much how you want it. Every time you change your variable names, you may have to update the tests.

## Story tables syntax

> Very basically, a **story table** contains a list of variables and values. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.

Author: Very good observation of missing information. Let me know if the below offers more clarity.

```suggestion
Very basically, a **story table** contains a list of variables that correspond to fields and the values you want for those fields. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.
```

> Very basically, a **story table** contains a list of variables and values. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.

Reviewer: Adding fields into the mix makes it more difficult to understand I think, mostly because you now have to distinguish between web fields (which these are) and PDF fields, and juggle 3 concepts around (fields, variables, and values) instead of just two (variables and values). The original might actually not be that bad now that I'm looking at it again. :shrug:

> Very basically, a **story table** contains a list of variables and values. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.

Author: How about this?

```suggestion
Very basically, a **story table** contains a list of variables and values. When the test is going through the interview and arrives on a page that has a field that defines the variable, the field will be set with the value. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions.
```

> Very basically, a **story table** contains a list of variables and values. The order in which you list the rows does not matter, which gives a lot more flexibility when developers are changing around the order of questions. Whenever pages with those variables appear, they will be given the value listed.

Reviewer: That's a great improvement! :+1:

> Tip: it can be easy to lose track of what variables you've included, so it can be useful keep the rows in alphabetical order.

Reviewer: ```suggestion
:::tip 
It can be easy to lose track of what variables you've included, so it can be useful keep the rows in alphabetical order.
:::
```

      | var | value | trigger |
```

This labels the three columns you will use. Under that, you add a row for every field that you want to interact with during the interview. Start with a blank row then fill it in:
```
      |  |  |  |
```

### var

In the **var** column, write the name of the variable you want to set **as it appears on question's screen**. Most times you can see that value in the YAML `question` block. Sometimes it will be generated by code in the `question` block. If you're not sure what name some code generates, you may have to ask the folks who wrote the code.

<!-- The Document Assembly Line library has built in questions that use such code.
- `users[i].name_fields()` generates all the name fields with the right index number (NOT 'i'). E.g. `users[0].name.first`, `users[0].name.suffix`, `users[1].name.first`, `users[1].name.suffix`
- `users[i].address_fields()` generates all the address fields with the right index number (NOT 'i'). E.g. `users[0].address.address`, `users[1].address.address` -->

### value

In the **value** column, write what you want the field to be set to. For checkboxes, `true` means 'checked' and `false` means 'unchecked'.

### trigger

> In the **trigger** column, write the name of the intrinsic name of the variable that will trigger the page for this field. This column is optional in most cases. In cases where the page is using an [index variable](https://docassemble.org/docs/fields.html#index%20variables) or a [`generic object`](https://docassemble.org/docs/modifiers.html#generic%20object).

Reviewer: ```suggestion
In the **trigger** column, write the name of the intrinsic name of the variable that will trigger the page for this field. This column is necessary in cases where the page is using an [index variable](https://docassemble.org/docs/fields.html#index%20variables) or a [`generic object`](https://docassemble.org/docs/modifiers.html#generic%20object), and optional everywhere else.
```

Author: I like it and I'll add one improvement where I see what I think is a typo of mine and another clarification:

```suggestion
In the **trigger** column, write the intrinsic name of the variable that will trigger the page where this field appears. This column is necessary in cases where the page is using an [index variable](https://docassemble.org/docs/fields.html#index%20variables) or a [`generic object`](https://docassemble.org/docs/modifiers.html#generic%20object), and optional everywhere else.
```

```
    When I set the variable "users[i].hair_color" to "blue"
```

Sign on a signature field. All signatures are the same.
```
    When I sign
```

Specifically for the Document Assembly Line 4-part name questions. No punctuation. We recommend you just use 2 names - the first name and last name - but you can have all these formats:

- Firstname Lastname
- Firstname Middlename Lastname
- Firstname Middlename Lastname Suffix (where suffix is one of the dropdown suffix choices, like `II`)

```
    When I set the name of "x[i]" to "Ulli User"
```

Specifically for the Document Assembly Line 5-field address questions. t can be any address you want that matches the format shown, commas and all.

```
    When I set the address of "users[0]" to "120 Tremont Street, Unit 1, Boston, MA 02108"
```

The Step for the story table, which is better described in sections above.
```
    I get to the question id "some yml block id" with this data:
```

### Other actions

```
    When I tap to continue

> ```

Reviewer: This one's duplicated?

Author: Yes. I wasn't sure whether to duplicate it or not - duplication is generally bad, but it's one I've used a lot under some circumstances, so I want it to be easy to find and it does seem to be to belong in both categories. What do you think?

Reviewer: That makes sense. IMO the design of this section on Steps might need some more work, but you're right, it's probably fine for now.

In docassemble, you can write `choices` a few different ways.

Bad

```
question: |
  What is your favorite color?
fields:
  - Colors: favorite_color
    input type: radio
    choices:
      - green
      - red
      - blue
```

Good

```
question: |
  What is your favorite color?
fields:
  - Colors: favorite_color
    input type: radio
    choices:
      - green: green
      - red: red
      - blue: blue
```

If your field is like the Bad example, the label of a choice is also its value and the translations will break the tests. This is because the label will be translated. Since the label written this way is also the value, the value in the translated version of the interview will be different than the value you wrote in for your test.

If your field is like the Good example, the value of the choice should never get translated so it should always be the same and the tests will be happy.

-->

## Test instructions details

### Add a new test

Go to your Playground > the dropdown Folders menu > Sources.

Add a new file that ends in the extension `.feature`. Ex: `has_children.feature`

> The next time you commit, your test will be run. You can add and edit multiple test files.

Reviewer: Is setting up the github action part of the first section, "Setup", that's commented out?

Author: Yes. In fact, the developer doesn't have to actively set up any actions. They just go through that setup interview. It should be self-explanatory when they run it, but I haven't user tested it yet. That said, the current PR is an MVP for making something to help conversations between the folks in the AL project about writing tests. I can do setup for our repos for now.

Let me know if we should discuss more. If this is a satisfactory answer, feel free to resolve this.

Reviewer: That's fine, skimmed right over that sentence, and it explains the part I thought was missing. Thanks!
